### PR TITLE
Suggested Code Cleanups

### PR DESCRIPTION
**Pull request by davidrevoledo**
## Description
Some suggested code clean ups.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x] If applicable, the public code is properly documented.
- [x] Pull request includes test coverage for the included changes.
- [x] The code builds without any errors.

### CODE REVIEW

File: src/Microsoft.Azure.Relay/Common/AsyncLock.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ class AsyncLock : IDisposable
 
         private readonly Task<LockRelease> lockRelease;
 
-        private bool disposed = false;
+        private bool disposed;
```
Comment (davidrevoledo) [inline src/Microsoft.Azure.Relay/Common/AsyncLock.cs]: default value

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 2]
```diff
@@ -173,7 +166,7 @@ public void Dispatch()
 
             if (outstandingReaders != null)
             {
-                ActionItem.Schedule(s => CompleteOutstandingReadersCallback(s), outstandingReaders);
+                ActionItem.Schedule(CompleteOutstandingReadersCallback, outstandingReaders);
```
Comment (davidrevoledo) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Call group method instead of lambda calling the method

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 3]
```diff
@@ -387,8 +380,8 @@ void EnqueueAndDispatch(Item item, bool canDispatchOnThisThread)
             bool disposeItem = false;
             IQueueReader reader = null;
             bool dispatchLater = false;
-            IQueueWaiter[] waiters = null;
-            bool itemAvailable = true;
+            IQueueWaiter[] waiters;
```
Comment (davidrevoledo) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: variable are assigned in all the execution paths

File: src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs [hunk 4]
```diff
@@ -133,22 +112,19 @@ public TimeSpan RemainingTime()
...
                 return this.originalTimeout;
             }
-            else if (this.deadline == DateTime.MaxValue)
+
+            if (this.deadline == DateTime.MaxValue)
             {
                 return TimeSpan.MaxValue;
             }
-            else
+
```
Comment (davidrevoledo) [inline src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs]: Else will never execute as the previous code do a return

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs (hunk 2 above)
Comment (dlstucki) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Your suggested version allocates every time the line runs. The existing version allocates and caches a delegate only the version first time. Subsequent calls do no allocate an additional delegates.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 5]
```diff
@@ -47,16 +47,9 @@ public int ReadersQueueCount
...
 
         // Users like ServiceModel can hook this abort ICommunicationObject or handle other non-IDisposable objects
-        public Action<T> DisposeItemCallback
-        {
-            get;
-            set;
-        }
+        public Action<T> DisposeItemCallback { get; set; }
 
-        object ThisLock
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: For internal reasons please don't use the `object ThisLock => this.itemQueue` syntax right now. (this code is mirrored in an internal repro.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 6]
```diff
@@ -343,11 +336,11 @@ static void CompleteWaitersLater(bool itemAvailable, IQueueWaiter[] waiters)
         {
             if (itemAvailable)
             {
-                ActionItem.Schedule(s => CompleteWaitersTrueCallback(s), waiters);
+                ActionItem.Schedule(CompleteWaitersTrueCallback, waiters);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Undo this change for fewer allocations as described above.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 7]
```diff
@@ -343,11 +336,11 @@ static void CompleteWaitersLater(bool itemAvailable, IQueueWaiter[] waiters)
...
             }
             else
             {
-                ActionItem.Schedule(s => CompleteWaitersFalseCallback(s), waiters);
+                ActionItem.Schedule(CompleteWaitersFalseCallback, waiters);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Undo this change for fewer allocations as described above.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 8]
```diff
@@ -365,7 +358,7 @@ static void InvokeDequeuedCallbackLater(Action dequeuedCallback)
         {
             if (dequeuedCallback != null)
             {
-                ActionItem.Schedule(s => OnInvokeDequeuedCallback(s), dequeuedCallback);
+                ActionItem.Schedule(OnInvokeDequeuedCallback, dequeuedCallback);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Undo this change for fewer allocations as described above.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 9]
```diff
@@ -447,7 +440,7 @@ void EnqueueAndDispatch(Item item, bool canDispatchOnThisThread)
 
             if (dispatchLater)
             {
-                ActionItem.Schedule(s => OnDispatchCallback(s), this);
+                ActionItem.Schedule(OnDispatchCallback, this);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Undo this change for fewer allocations as described above.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 10]
```diff
@@ -470,11 +463,9 @@ bool EnqueueWithoutDispatch(Item item)
...
                         return false;
                     }
-                    else
-                    {
-                        itemQueue.EnqueuePendingItem(item);
-                        return true;
-                    }
+
+                    itemQueue.EnqueuePendingItem(item);
+                    return true;
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: This is really just a minor style preference one way or another. Though less indentation is generally good IMO so it's fine.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 11]
```diff
@@ -603,7 +585,7 @@ public AsyncQueueReader(InputQueue<T> inputQueue, CancellationToken cancellation
                 : base(state)
             {
                 this.inputQueue = inputQueue;
-                this.cancelRegistration = cancellationToken.Register(s => CancelCallback(s), this);
+                this.cancelRegistration = cancellationToken.Register(CancelCallback, this);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Undo this change for fewer allocations as mentioned above. (Yes I know VS2017 or some plugins try to correct this but I want fewer allocations, not slightly less code).

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 12]
```diff
@@ -638,7 +620,7 @@ class AsyncQueueWaiter : TaskCompletionSource<bool>, IQueueWaiter
             public AsyncQueueWaiter(CancellationToken cancellationToken, object state)
                 : base(state)
             {
-                this.cancelRegistration = cancellationToken.Register(s => CancelCallback(s), this);
+                this.cancelRegistration = cancellationToken.Register(CancelCallback, this);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Undo this change for fewer allocations as mentioned above. (Yes I know VS2017 or some plugins try to correct this but I want fewer allocations, not slightly less code).

File: src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs [hunk 13]
```diff
@@ -12,7 +12,7 @@ struct TimeoutHelper
     {
         DateTime deadline;
         bool deadlineSet;
-        TimeSpan originalTimeout;
+        readonly TimeSpan originalTimeout;
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs]: Place readonly fields before mutable fields.

1. static readonly field
2. static field
3. readonly field
4. field

File: src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs [hunk 14]
```diff
@@ -133,22 +112,19 @@ public TimeSpan RemainingTime()
                 this.SetDeadline();
                 return this.originalTimeout;
             }
-            else if (this.deadline == DateTime.MaxValue)
+
+            if (this.deadline == DateTime.MaxValue)
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs]: Please change this back to else if to avoid requiring the empty line between the if and this block in question. (fewer empty lines are generally preferable.)

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs (hunk 2 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Ok I wasn't aware of that, will investigate and remove it. thks @dlstucki

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs (hunk 5 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Ok will keep in mind for future contributions :) I'm starting here and want to help as much I can :D

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs (hunk 6 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Sure

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs (hunk 10 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: yes I think is good if  {}  if {} or if {}  return. In others repos I've contributed the team tried to avoid `else` specifically

File: src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs (hunk 13 above)
Comment (davidrevoledo) [reply src/Microsoft.Azure.Relay/Common/TimeoutHelper.cs]: 💯

**Comment by davidrevoledo** [comment]
@dlstucki Thanks for your feedback this is my first contribution here, and I'm not aware of the code guidelines here. Will keep in mind all of them.

My idea is start contributing here as much I can :)

Thanks for your feedback much appreciated 💯

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs (hunk 5 above)
Comment (dlstucki) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: If this is related to my comment about not using `object ThisLock => this.itemQueue;` I retracted those statements. The newer `Property => field;` syntax looks like it's ok in the internal mirror repo. However, code churn just for the sake of updating to new language features when it has no other benefit is a low priority.

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs (hunk 2 above)
Comment (dlstucki) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: I made a sample to show what I'm talking about:
``` C#
using System;
using System.Diagnostics;

class Program
{
    static void Main(string[] args)
    {
        const long Iterations = 2 * 1000 * 1000 * 1000L;
        var program = new Program();
        var stopwatch = new Stopwatch();
        stopwatch.Reset();
        GC.KeepAlive(stopwatch.Elapsed);

        Console.WriteLine("Example1");
        stopwatch.Restart();
        for (long i = 0; i < Iterations; i++)
        {
            ActionItemSchedule(s => CompleteWaitersCallback(s), program);
        }
        Console.WriteLine($"Example1 completed in {stopwatch.Elapsed}");

        Console.WriteLine("Example2");
        stopwatch.Restart();
        for (long i = 0; i < Iterations; i++)
        {
            ActionItemSchedule(CompleteWaitersCallback, program);
        }
        Console.WriteLine($"Example2 completed in {stopwatch.Elapsed}");
    }

    static void ActionItemSchedule(Action<object> action, object state)
    {
        action(state);
    }

    static void CompleteWaitersCallback(object state)
    {
        GC.KeepAlive(state);
    }
}
```
Sample output on my machine:
> Example1
> Example1 completed in 00:00:09.7257016
> Example2
> Example2 completed in 00:00:25.8970678

Also, the perf counters show a big difference:
image

File: src/Microsoft.Azure.Relay/Common/InputQueue.cs [hunk 15]
```diff
@@ -568,20 +559,11 @@ public Item(Exception exception, Action dequeuedCallback)
                 this.dequeuedCallback = dequeuedCallback;
             }
 
-            public Action DequeuedCallback
-            {
-                get { return this.dequeuedCallback; }
-            }
+            public Action DequeuedCallback => this.dequeuedCallback;
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Common/InputQueue.cs]: As long as you're touching these could we eliminate the fields altogether and just use the readonly auto-property syntax?
``` C#
public Action DequeuedCallback { get; }

public Exception Exception { get; }

public Exception Exception { get; }
```

Comment (davidrevoledo) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Sure Personally  I prefer that sintaxis 

`public Action DequeuedCallback { get; }` if is readonly and 
`public Action DequeuedCallback { get; private set; }` if need to be modified outside the ctor

Comment (dlstucki) [reply src/Microsoft.Azure.Relay/Common/InputQueue.cs]: Hopefully none of those fields/properties are modified outside the .ctor.

**Comment by dlstucki** [comment]
I expect you'll add a new commit addressing the comments, then we can probably merge the PR.

**Comment by davidrevoledo** [comment]
@dlstucki done.

**Comment by davidrevoledo** [comment]
@dlstucki thks a lot, please feel free to keep in mind for me for up-for-grabs tasks or anything I'll investige more in deep the library as I was using relay so far as WCF communication and planning to implement something with this :)
